File: src/wipe.c
#include "wipe.h"

// write methods
unsigned char _write0       (void) { return 0b00000000; }
unsigned char _write1       (void) { return 0b11111111; }
unsigned char _write01      (void) { return 0b01010101; }
unsigned char _writeRandom  (void) { return rand();     } // NOLINT(cert-msc50-cpp)

// check methods
int _checkNone  (unsigned char byte) { return 1;                  }
int _check0     (unsigned char byte) { return byte == 0b00000000; }
int _check1     (unsigned char byte) { return byte == 0b11111111; }
int _check01    (unsigned char byte) { return byte == 0b01010101; }

// wipe method
void wipeFiles(char paths[][MAX_PATH_LENGTH], size_t pathsLength, algorithm algorithm) {
    printf("Starting the wiping sequence of %zu file%s with method : %s\n", pathsLength, pathsLength > 1 ? "s" : "", algorithm.name);
    // for each path
    for (size_t i = 0; i < pathsLength; i++) {
        // check if is writable again by security
        if(access(paths[i], W_OK) == 0) {
            if(verboseFlag) printf(CLEAR_LINE "Opening file \"%s\"\n", paths[i]);
            // Open file stream (r+b ==> read + write at the top of the file in binary)
            FILE *fp = fopen(paths[i], "r+b");
            if(fp == NULL) {
                // If an error as occurred while opening the file
                if(verboseFlag) printf(CLEAR_LINE COLOR_RED "Cannot open file \"%s\"" COLOR_RESET, paths[i]);
            } else {
                // get file size
                fseek(fp, 0, SEEK_END);
                unsigned long int fileSize = ftell(fp);
                unsigned long int fileSizeDouble = fileSize * 2; // The double will be in use for the verification sequence

                if(verboseFlag) printf(CLEAR_LINE "File size: %zu bytes\n", fileSize);

                // variables
                int passCounter;
                unsigned long int offsetCounter;
                int lastPercent = 0;
                int percent = 0;
                int errorOnWriteMethod = 0;
                unsigned char byte;

                // for each pass sequence
                for(passCounter = 0; passCounter < algorithm.size; passCounter++) {
                    if(verboseFlag) printf(CLEAR_LINE "Starting pass %d of %d\n", passCounter + 1, algorithm.size);

                    // return on the top of the file
                    fseek(fp, 0, SEEK_SET);

                    // for each byte of the file
                    for(offsetCounter = 1; offsetCounter <= fileSizeDouble; offsetCounter++) {
                        // Calculate the percentage of advancement and convert it to int instead of float
                        percent = (int) roundf(((((float) offsetCounter / (float) fileSizeDouble) * ((float) passCounter + 1))/(float) algorithm.size) * 100.0f);

                        // If the percentage is higher that the previous one
                        if(percent > lastPercent) {
                            lastPercent = percent;
                            printf(CLEAR_LINE "[file %zu/%zu] [%3d%%] Wiping file: Pass %d of %d", i + 1, pathsLength, percent, passCounter + 1, algorithm.size);
                            fflush(stdout);
                        }


                        if(offsetCounter <= fileSize) { // If it's under the half of the iterator (writing part)
                            // write a byte to the file and move cursor to the next one
                            fprintf(fp, "%c", algorithm.writes[passCounter]());

                            // if write is finished, then return to the top of the file for the verification
                            if(offsetCounter == fileSize) {
                                // return on the top of the file
                                fseek(fp, 0, SEEK_SET);

                                if(verboseFlag) printf(CLEAR_LINE "End of writing for this pass, now place for verification\n");
                            }
                        } else { // If it's after the half of the iterator (verification part)
                            fread(&byte, 1, 1, fp);
                            if(algorithm.checks[passCounter](byte) != 1 && errorOnWriteMethod == 0) {
                                errorOnWriteMethod = 1;
                                printf(CLEAR_LINE "An error has occurred while wiping the file \"%s\"\nOne or more bytes were not rewritten as they should have been.\nPlease verify that the file is editable and that it is not being used by another process.\n", paths[i]);
                                break;
                            }
                        }
                    }
                }

                printf(CLEAR_LINE "[file %zu/%zu] [100%%] Wiping file: Pass %d of %d\n", i + 1, pathsLength, passCounter, algorithm.size);
                fflush(stdout);

                if(verboseFlag) printf("End for this file\n");

                // Close the file stream
                if(fclose(fp) == 0) {
                    if(verboseFlag) printf("Successfully close the stream\n");
                } else {
                    if(verboseFlag) printf("Cannot close the stream\n");
                }

                // If the `--keep` parameter as not been passed
                if(keepFilesAfterOverwriting == 0) {
                    // Try to remove the file
                    if(remove(paths[i]) == 0) {
                        if(verboseFlag) printf("Successfully delete the file from your system\n");
                    } else {
                        printf("An error has occurred, cannot remove from your system the file \"%s\"\nPlease verify that the file is editable and that it is not being used by another process.\n", paths[i]);
                    }
                }
            }
        } else {
            if(verboseFlag) printf(COLOR_RED "Cannot write file \"%s\" while it should be" COLOR_RESET, paths[i]);
        }
    }
}

// A function that is passed to `nftw`to remove all files in a directory then delete the directory
int wipeDir(const char *fpath, const struct stat *sb, int tflag, struct FTW *ftwbuf) {
    if(S_ISREG(sb->st_mode) && access(fpath, W_OK) == 0) { // If it's a writable regular file
        if(verboseFlag) printf("Found a writable file : %s\n", fpath);

        if(remove(fpath) == 0) {
            if(verboseFlag) printf("Successfully deleted it\n");
        } else {
            if(verboseFlag) printf("Cannot delete it\n");
        }
    } else if(S_ISDIR(sb->st_mode)) { // If it's a directory
        if(verboseFlag) printf("Found a directory : %s\n", fpath);

        if(rmdir(fpath) == 0) {
            if(verboseFlag) printf("Successfully deleted it\n");
        } else {
            if(verboseFlag) printf("Cannot delete it. It may not be empty\n");
        }
    } else if(S_ISLNK(sb->st_mode)) { // If it's a symbolic link
        if(verboseFlag) printf("Found a symbolic link : %s\n", fpath);

        if(unlink(fpath) == 0) {
            if(verboseFlag) printf("Successfully unlinked it\n");
        } else {
            if(verboseFlag) printf("Cannot unlinked it\n");
        }
    }

    return 0;
}

File: src/wipe.rs
use positioned_io::{WriteBytesAtExt, ReadBytesAtExt};
use std::fs::{File, OpenOptions, remove_file};
use std::process::exit;
use std::io::{Error, stdout, Write};
use crate::algorithms::ALGORITHMS;

// Algorithm struct
pub struct Algorithm {
    pub name: &'static str,
    pub(crate) writes: &'static [fn() -> u8],
    pub(crate) checks: &'static [fn(u8) -> bool],
}

static CLEAR_LINE: &'static str = "\x1b[2K\r";

pub fn wipeFiles(files: Vec<String>, algorithmNbr: u16, keep: bool, verbose: bool) {
    // Check if the algorithm number is valid
    if algorithmNbr < 1 || algorithmNbr > ALGORITHMS.len() as u16 {
        println!("Invalid algorithm number, please use the --list option to see the available algorithms");
        exit(1);
    }

    // Get the algorithm
    let algorithm: &Algorithm = &ALGORITHMS[(algorithmNbr - 1) as usize];

    // Print the algorithm name if verbose mode is enabled
    if verbose {
        println!("Using algorithm: {}", algorithm.name);
        println!("");
    }

    // For each file
    let totalFiles: u64 = files.len() as u64;
    for (index, file) in files.iter().enumerate() {
        let index: u64 = (index + 1) as u64;
        if crate::utils::isRegularFile(file.clone()) {
            if crate::utils::isWritable(file.clone()) {
                print!("Opening {} ...", file);
                stdout().flush().ok();

                // Open the file stream
                let fp: Result<File, Error> = OpenOptions::new().read(true).write(true).open(file.clone());

                // If the file is open
                if fp.is_ok() {
                    let mut fp: File = fp.unwrap();
                    let fileSize: u64 = fp.metadata().unwrap().len();
                    let mut offset: u64;
                    let mut percent: f32 = 0.00;
                    let mut writeError: bool = false;
                    let mut checkError: bool = false;

                    // Print the percent
                    print!("{}[file {}/{}][  0.00%] - {}", CLEAR_LINE, index, totalFiles, file);
                    stdout().flush().ok();

                    // For each write function
                    for i in 0..algorithm.writes.len() as u64 {
                        // Reset offset
                        offset = 0;

                        // For each byte write
                        for ii in 0..fileSize {
                            // Calculate the percent
                            let mut newPercent: f32 = 100.0 * (((i * fileSize * 2) + ii) as f32) / (algorithm.writes.len() as u64 * fileSize * 2) as f32;
                            newPercent = f32::trunc(newPercent  * 100.0) / 100.0;

                            // Print the percent
                            if newPercent != percent {
                                percent = newPercent;
                                print!("{}[file {}/{}][{: >6.2}%] - {}", CLEAR_LINE, index, totalFiles, percent, file);
                                stdout().flush().ok();
                            }

                            // Write the byte
                            let writeResult: Result<(), Error> = fp.write_u8_at(offset, algorithm.writes[i as usize]());
                            if writeResult.is_err() {
                                writeError = true;
                                println!("{}An error as occured while writing {} :", CLEAR_LINE, file);
                                println!("\t{}", writeResult.err().unwrap());
                                break;
                            }

                            // Set offset to the next position
                            offset += 1;
                        }

                        // Print the percent
                        //print!("{}[{}] - {}", CLEAR_LINE, if checkError { " ERROR " } else { " 50.00%" }, file);
                        //stdout().flush().ok();

                        // Reset offset
                        offset = 0;

                        // For each byte check
                        for ii in 0..fileSize {
                            // If an error has occured while writting skip
                            if writeError {
                                break;
                            }

                            // Calculate the percent
                            let mut newPercent: f32 = 100.0 * (((i * fileSize * 2) + ii) as f32 + fileSize as f32) / (algorithm.writes.len() as u64 * fileSize * 2) as f32;
                            newPercent = f32::trunc(newPercent  * 100.0) / 100.0;

                            // Print the percent
                            if newPercent != percent {
                                percent = newPercent;
                                print!("{}[file {}/{}][{: >6.2}%] - {}", CLEAR_LINE, index, totalFiles, percent, file);
                                stdout().flush().ok();
                            }

                            // Read the byte
                            let readResult: Result<u8, Error> = fp.read_u8_at(offset);

                            // Check the byte
                            if readResult.is_err() {
                                checkError = true;
                                println!("{}An error as occured while checking {} :", CLEAR_LINE, file);
                                println!("\t{}", readResult.err().unwrap());
                                break;
                            } else {
                                if !algorithm.checks[i as usize](readResult.unwrap()) {
                                    checkError = true;
                                    println!("{}An error as occured while checking {} :", CLEAR_LINE, file);
                                    println!("\tThe byte at offset {} is not the expected one", offset);
                                    break;
                                }
                            }

                            // Set offset to the next position
                            offset += 1;
                        }

                        // Print the percent
                        print!("{}[file {}/{}][{}] - {}", CLEAR_LINE, index, totalFiles, if checkError { " ERROR " } else { "100.00%" }, file);
                        stdout().flush().ok();
                    }

                    // Delete the file if keep is false
                    if !keep {
                        if verbose {
                            print!("\nDeleting {} ...", file);
                        }
                        let deleteResult: Result<(), Error> = remove_file(file.clone());
                        if deleteResult.is_err() {
                            println!("\nAn error as occured while deleting {} :", file);
                            println!("\t{}", deleteResult.err().unwrap());
                        }
                    }

                    println!("");
                    stdout().flush().ok();

                } else {
                    println!("{} is not readable, skipping...", file);
                }
            } else {
                println!("{} is not writable, skipping...", file);
            }
        } else {
            println!("{} is not a regular file, skipping...", file);
        }
    }
}